core/src/main/java/com/graphhopper/routing/ch/NodeBasedNodeContractor.java

```diff
@@ -30,7 +33,9 @@
 class NodeBasedNodeContractor implements NodeContractor {
     private final CHPreparationGraph prepareGraph;
     private final Params params = new Params();
-    private ShortcutHandler shortcutHandler;
+    // todonow: maybe use a set to prevent duplicates instead?
+    private List<Shortcut> shortcuts = new ArrayList<>();
```

Comment:
A HashSet is more memory intense as the underlying array is 30%-60% empty but might be indeed faster. We would have to try this :)

Reply:
This todo has also been there for some time :) I just moved it here from another place.

Reply:
ah, I thought you changed it from "todo" to "todonow"?

Reply:
Ah very true :) I probably thought it is time to try this and fix the todo. I ran into many things that I thought aren't ideal and it is hard to keep the focus on this specific refactoring. For now I'm happy all tests are green and `CHGraphImpl` is gone, but now I need to clean up more.